packages/core/src/api.ts

```diff
@@ -175,15 +194,14 @@ export const deploy = async (
   }
 
   const postDeployAction = async (appliedChanges: ReadonlyArray<Change>): Promise<void> => {
-    await promises.array.series(appliedChanges.map(change => async () => {
-      const updatedElement = await getUpdatedElement(change)
-      if (change.action === 'remove' && !isFieldChange(change)) {
-        await workspace.state().remove(updatedElement.elemID)
-      } else {
-        await workspace.state().set(updatedElement)
+    const detailedChanges = appliedChanges.map(change => getDetailedChangesFromChanges(change)).flat()
+    await workspace.state().updateStateFromChanges({ changes: detailedChanges })
+
+    await awu(appliedChanges).filter(change => (isAdditionOrModificationChange(change) || isFieldChange(change)))
+      .forEach(async change => {
+        const updatedElement = await getUpdatedElement(change)
```

Comment:
by updating the state first you are changing the behavior of `getUpdatedElement` (because in some cases it gets the element from the state).
not sure if it actually breaks anything (the fact that we are "double applying" a change) but it would be very confusing to follow for future readers